yapapi/engine.py

```diff
@@ -229,53 +232,65 @@ def subnet_tag(self) -> Optional[str]:
         """Return the name of the subnet used by this engine, or `None` if it is not set."""
         return self._subnet
 
+    @property
+    def operative(self) -> bool:
+        return self._operative
+
     def emit(self, event: events.Event) -> None:
         """Emit an event to be consumed by this engine's event consumer."""
         if self._wrapped_consumer:
             self._wrapped_consumer.async_call(event)
 
     async def __aenter__(self) -> "_Engine":
-        """Initialize resources and start background services used by this engine."""
-
         try:
-            stack = self._stack
+            await self._start()
+            return self
+        except:
+            await self._stop(*sys.exc_info())
+            raise
 
-            await stack.enter_async_context(self._wrapped_consumer)
+    async def __aexit__(self, *exc_info) -> Optional[bool]:
+        return await self._stop(*exc_info)
 
-            def report_shutdown(*exc_info):
-                if any(item for item in exc_info):
-                    self.emit(events.ShutdownFinished(exc_info=exc_info))  # noqa
-                else:
-                    self.emit(events.ShutdownFinished())
+    async def _stop(self, *exc_info) -> Optional[bool]:
+        self._operative = False
+        return await self._stack.__aexit__(*exc_info)
 
-            stack.push(report_shutdown)
+    async def _start(self) -> None:
+        self._operative = True
```

Comment:
Shouldn't we set this to `True` at the end of the method? If any error occurs before the end of initialization we may end up with a `Golem` instance that is `operative` but not properly initialized.

Reply:
My reasoning was that we have to guard against more than one `start` running at the same time, as in:
```
golem = Golem(...)
loop.create_task(golem.start)
loop.create_task(golem.start)
```
(I don't think anyone will write **exactly this** but this might somehow happen in a more complex code, I guess?).

We could store two bits of information, `start_started`, `start_ended`.

But maybe we should rename`operative`? It is now if fact `started_not_stopped`. Or maybe just remove a public property?

Reply:
You've raised a good point. Consider again your example:
```
golem = Golem(...)
loop.create_task(golem.start())   # 1
loop.create_task(golem.start())   # 2
```
Suppose `golem.start()` in line `#1` is in progress and `golem.start()` in `#2` starts. We don't want the second instance of `golem._start()` to run concurrently with the first one, and your current solution will prevent that. But we also don't want `golem.start()` in line `#2` exiting immediately, as if `golem` were already properly started, while `_start()` is still in progress! This is even more important when we replace both `golem.start()`'s with `golem.execute_tasks()`. What we do want is for the second `golem.start()` to _wait_ until the first one finishes. So we need some kind of task synchronisation mechanism, e.g. an `asyncio.Lock`.

Reply:
That's a really good point. I need to think about this : )